File: src/monid_finance_mcp/providers/us/web_extract.py
"""Context.dev /web/extract request and response handling.

The validated route: POST with a caller-supplied JSON Schema,
factCheck true, maxDepth 0 (only the starting page). The output
envelope is ``{status, url, urls_analyzed, data}`` where ``data``
matches the supplied schema and unsupported fields come back null.
"""
from __future__ import annotations

from monid_finance_mcp.models import JsonObject, JsonValue
from monid_finance_mcp.providers.us.normalize import SchemaDriftError

EXTRACT_ENDPOINT = "/web/extract"


def extract_request(
    *,
    url: str,
    schema: JsonObject,
    instructions: str,
    max_pages: int = 1,
) -> JsonObject:
    """Build the /web/extract request body for one page, fact-checked."""
    return {
        "url": url,
        "schema": schema,
        "instructions": instructions,
        "factCheck": True,
        "maxPages": max_pages,
        "maxDepth": 0,
    }


def parse_extract_output(value: JsonValue, *, expected_url: str) -> JsonObject:
    """Return the extracted ``data`` object, validating the envelope."""
    if not isinstance(value, dict):
        raise SchemaDriftError("Context.dev extract payload must be an object")
    payload = value
    if isinstance(payload.get("data"), dict) and "status" in payload:
        if payload.get("status") != "ok":
            raise SchemaDriftError("Context.dev extract did not report status ok")
        data = payload.get("data")
        if not isinstance(data, dict):
            raise SchemaDriftError("Context.dev extract data must be an object")
        return data
    del expected_url, payload
    raise SchemaDriftError("Context.dev extract envelope is missing status or data")

File: go/service/webextract.go
// This file ports web_extract.py: the Context.dev /web/extract request
// body and response-envelope handling shared by get_segmented_financials
// and the three get_kpi_* tools.
package service

import (
	"encoding/json"

	"github.com/belazy/monid-finance/providers"
)

// extractEndpoint mirrors web_extract.EXTRACT_ENDPOINT.
const extractEndpoint = "/web/extract"

// extractRequestBody mirrors web_extract.extract_request: one page,
// fact-checked, with the caller-supplied JSON Schema and instructions.
func extractRequestBody(url string, schema map[string]any, instructions string) map[string]any {
	return map[string]any{
		"url":          url,
		"schema":       schema,
		"instructions": instructions,
		"factCheck":    true,
		"maxPages":     1,
		"maxDepth":     0,
	}
}

// parseExtractOutput mirrors web_extract.parse_extract_output: validates
// the {status, url, urls_analyzed, data} envelope and returns the
// extracted data object.
func parseExtractOutput(raw json.RawMessage) (map[string]any, error) {
	var value any
	if err := json.Unmarshal(raw, &value); err != nil {
		return nil, &providers.SchemaDriftError{Msg: "Context.dev extract payload must be an object"}
	}
	payload, ok := value.(map[string]any)
	if !ok {
		return nil, &providers.SchemaDriftError{Msg: "Context.dev extract payload must be an object"}
	}
	data, hasData := payload["data"].(map[string]any)
	_, hasStatus := payload["status"]
	if !hasData || !hasStatus {
		return nil, &providers.SchemaDriftError{Msg: "Context.dev extract envelope is missing status or data"}
	}
	if status, _ := payload["status"].(string); status != "ok" {
		return nil, &providers.SchemaDriftError{Msg: "Context.dev extract did not report status ok"}
	}
	return data, nil
}
